rust/fury/src/util.rs

```diff
@@ -0,0 +1,90 @@
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Swapping the high 8 bits and the low 8 bits of a 16-bit value
+fn swap_endian(value: u16) -> u16 {
+    (value << 8) | (value >> 8)
+}
+
+pub fn to_utf8(utf16: &[u16], is_little_endian: bool) -> Result<Vec<u8>, String> {
+    // Pre-allocating approximate capacity to minimize dynamic resizing
+    let mut utf8_bytes = Vec::with_capacity(utf16.len() * 2);
+    let mut iter = utf16.iter();
+    while let Some(&wc) = iter.next() {
+        let wc = if is_little_endian {
+            swap_endian(wc)
+        } else {
+            wc
+        };
+        match wc {
+            // 1-byte UTF-8
+            // [0000|0000|0ccc|cccc] => [0ccc|cccc]
+            code_point if code_point < 0x80 => {
+                utf8_bytes.push(code_point as u8);
+            }
+            // 2-byte UTF-8
+            code_point if code_point < 0x800 => {
+                // [0000|0bbb|bbcc|cccc] => [110|bbbbb], [10|cccccc]
+                let second = ((code_point & 0b11_1111) as u8) | 0b1000_0000;
+                let first = ((code_point >> 6 & 0b1_1111) as u8) | 0b1100_0000;
+                utf8_bytes.push(first);
+                utf8_bytes.push(second);
+            }
+            // Surrogate pair (4-byte UTF-8)
+            wc1 if (0xd800..=0xdbff).contains(&wc1) => {
+                // Need extra byte
+                if let Some(&wc2) = iter.next() {
+                    let wc2 = if is_little_endian {
+                        swap_endian(wc2)
+                    } else {
+                        wc2
+                    };
+                    if !(0xdc00..=0xdfff).contains(&wc2) {
+                        return Err("Invalid UTF-16 string: wrong surrogate pair".to_string());
+                    }
+                    // utf16 to unicode
+                    let code_point =
+                        ((((wc1 as u32) - 0xd800) << 10) | ((wc2 as u32) - 0xdc00)) + 0x10000;
+                    // 11110??? 10?????? 10?????? 10??????
+                    // Need 21 bit suffix of code_point
+                    let fourth = ((code_point & 0b11_1111) as u8) | 0b1000_0000;
+                    let third = ((code_point >> 6 & 0b11_1111) as u8) | 0b1000_0000;
+                    let second = ((code_point >> 12 & 0b11_1111) as u8) | 0b1000_0000;
+                    let first = ((code_point >> 18 & 0b111) as u8) | 0b1111_0000;
+                    utf8_bytes.push(first);
```

Comment:
`push` will check the memory boundary, since we initialize utf8_bytes with a fixed capacity, we can avoid this overhead by using a more efficient way. Merging four u8 values into a single u32 to reduce additional memory writes.

Reply:
Thanks for your advice. I have replaced `push` with `ptr.add(offset).write(value)` inside an `unsafe` block. However, I can't write a `u32` into `Vec<u8>` at once, so I'm still using multiple write operations.

Reply:
You can cast u8 ptr to u32 ptr.  for instance  `buf.as_mut_ptr() as *mut i32;`

Reply:
I found that write u32 to [u8:4] will be affected by processor's endianness.
```rust
fn main() {
    let mut vec = vec![0u8; 4];
    unsafe {
        let ptr = vec.as_mut_ptr() as *mut u32;
        ptr.write(0xa1b2c3d4);
        vec.set_len(4);
    }
    for i in 0..vec.len() {
        print!("0x{:02x} ", vec[i]);
    }
    // print: 0xd4 0xc3 0xb2 0xa1
}
````
So I use `#[cfg(target_endian = "little | big")]` to handle this.
```rust
#[cfg(target_endian = "little")]
ptr.write(0xd4c3b2a1);

#[cfg(target_endian = "big")]
ptr.write(0xa1b2c3d4);
```